Framework/Parallel/src/IO/EventParser.cpp

```diff
@@ -0,0 +1,284 @@
+
+  for (; pulse < eventIndex.size(); ++pulse) {
+    size_t icount =
+        (pulse != eventIndex.size() - 1 ? eventIndex[pulse + 1] : rangeEnd) -
+        eventIndex[pulse];
+    if (rangeStart >= static_cast<size_t>(eventIndex[pulse]) &&
+        rangeStart < static_cast<size_t>(eventIndex[pulse]) + icount)
+      startPulse = pulse;
+    if (rangeEnd > static_cast<size_t>(eventIndex[pulse]) &&
+        rangeEnd <= static_cast<size_t>(eventIndex[pulse]) + icount)
+      endPulse = pulse + 1;
+  }
+
+  curr = endPulse;
+
+  return {startPulse, endPulse};
+}
+
+/** Extracts event information from the list of time offsets and global spectrum
+ * indices using the event_index and event_time_offset tables provided from
+ * file. These events are separated according to MPI ranks.
+ * @param rankData vector which stores vectors of data for each mpi rank.
+ * @param globalSpectrumIndex list of spectrum indices corresponding to tof data
+ * @param eventTimeOffset tof data
+ * @param offset File offset (index) for tof data. Used to track event_index and
+ * event_time_zero positions
+ */
+template <class IndexType, class TimeZeroType, class TimeOffsetType>
+void EventParser<IndexType, TimeZeroType, TimeOffsetType>::
+    extractEventsForRanks(std::vector<std::vector<Event>> &rankData,
+                          const int32_t *globalSpectrumIndex,
+                          const TimeOffsetType *eventTimeOffset,
+                          const LoadRange &range) {
+  for (auto &item : rankData)
+    item.clear();
+
+  auto offset = range.eventOffset;
+  auto count = range.eventCount;
+  // Determine start and end pulse/s for chunk of data provided.
+  auto result = findStartAndEndPulseIndices(m_eventIndex, offset, count,
+                                            m_posInEventIndex);
+
+  for (size_t pulse = result.first; pulse <= result.second; ++pulse) {
+    const auto start =
+        std::max(
+            (pulse == 0 ? 0 : static_cast<size_t>(m_eventIndex[pulse - 1])),
+            offset) -
+        offset;
+    const auto end = std::min(static_cast<size_t>(pulse == m_eventIndex.size()
+                                                      ? offset + count
+                                                      : m_eventIndex[pulse]),
+                              offset + count) -
+                     offset;
+
+    for (size_t i = static_cast<size_t>(start); i < static_cast<size_t>(end);
+         ++i) {
+      int rank = globalSpectrumIndex[i] % m_comm.size();
+      rankData[rank].push_back(
+          Event{globalSpectrumIndex[i],
+                TofEvent{static_cast<double>(eventTimeOffset[i]),
+                         m_eventTimeZero[pulse]}});
+    }
+  }
+}
+
+/** Uses MPI calls to redistribute chunks which must be processed on certain
+ *ranks.
+ *@param result output data which must be processed on current rank. (May
+ *be updated by other ranks)
+ *@param data Data on this rank which belongs to several other ranks.
```

Comment:
If this `data` param represents data on other ranks, presumably we need to indicate that it cannot be modified on the current rank. i.e. should be ```const std::vector<std::vector<Event>>&```?